src/Sarif.Converters/FortifyFprConverter.cs

```diff
@@ -0,0 +1,631 @@
+
+        private string FprTypeToMimeType(string fileType)
+        {
+            switch (fileType)
+            {
+                case "xml":
+                    return "text/xml";
+
+                case "tsql":
+                    return "text/x-sql";
+
+                default:
+                    return "unknown/" + fileType;
+            }
+        }
+
+        private void ParseVulnerabilities()
+        {
+            _reader.Read();
+            while (!AtEndOf(_strings.Vulnerabilities))
+            {
+                if (AtStartOfNonEmpty(_strings.Vulnerability))
+                {
+                    ParseVulnerability();
+                }
+                else
+                {
+                    _reader.Read();
+                }
+            }
+        }
+
+        private void ParseVulnerability()
+        {
+            var result = new Result
+            {
+                Locations = new List<Location>(),
+                CodeFlows = new List<CodeFlow>
+                {
+                    new CodeFlow
+                    {
+                        Locations = new List<AnnotatedCodeLocation>()
+                    }
+                }
+            };
+
+            _reader.Read();
+            while (!AtEndOf(_strings.Vulnerability))
+            {
+                if (AtStartOfNonEmpty(_strings.ClassId))
+                {
+                    result.RuleId = _reader.ReadElementContentAsString();
+                }
+                else if (AtStartOfNonEmpty(_strings.AnalysisInfo))
+                {
+                    ParseLocationFromAnalysisInfo(result);
+                }
+
+                _reader.Read();
+            }
+
+            _results.Add(result);
+        }
+
+        private void ParseLocationFromAnalysisInfo(Result result)
+        {
+            CodeFlow codeFlow = result.CodeFlows.First();
+            int step = 0;
+
+            _reader.Read();
+            string lastSnippetId = null;
+            while (!AtEndOf(_strings.AnalysisInfo))
+            {
+                // Note: SourceLocation is an empty element (it has only attributes),
+                // so we can't call AtStartOfNonEmpty here.
+                if (AtStartOf(_strings.SourceLocation))
+                {
+                    string snippetId = _reader.GetAttribute(_strings.SnippetAttribute);
+                    PhysicalLocation physLoc = ParsePhysicalLocationFromSourceInfo();
+
+                    var acl = new AnnotatedCodeLocation
+                    {
+                        Step = step++,
+                        PhysicalLocation = physLoc
+                    };
+
+                    // Remember the id of the snippet associated with this location.
+                    // We'll use it to fill the snippet text when we read the Snippets element later on.
+                    if (!string.IsNullOrEmpty(snippetId))
+                    {
+                        _aclToSnippetIdDictionary.Add(acl, snippetId);
```

Comment:
No, I didn't think about the subtle difference between `d[key] = val` and `d.Add(key, val)` (namely, that the latter throws if `key` is already present in `d`). In this case, since `acl` is a newly created object, and since all the SARIF classes have reference equality semantics, it's guaranteed not to exist in the dictionary, so it doesn't matter which I use.

---
In reply to: [94535642](https://github.com/Microsoft/sarif-sdk/pull/554#discussion_r94535642) [](ancestors = 94535642)